HARK/distribution.py

```diff
@@ -930,29 +929,288 @@ def draw(self, N, X=None, exact_match=False):
+            f_query = self.X
+        else:
+            # if a function is provided, we need to add one more dimension,
+            # the nature dimension, to any inputs that are n-dim arrays.
+            # This allows numpy to easily broadcast the function's output.
+            # For more information on broadcasting, see:
+            # https://numpy.org/doc/stable/user/basics.broadcasting.html#general-broadcasting-rules
+            args = [
+                arg[..., np.newaxis] if isinstance(arg, np.ndarray) else arg
+                for arg in args
+            ]
+
+            f_query = func(self.X, *args)
+
+        f_exp = np.dot(f_query, self.pmf)
+
+        return f_exp
+
+    def dist_of_func(self, func=lambda x: x, xarray=False, *args, **kwargs):
+        """
+        Finds the distribution of a random variable Y that is a function
+        of discrete random variable X, Y=f(X).
+
+        Parameters
+        ----------
+        func : function
+            The function to be evaluated.
+            This function should take the full array of distribution values.
+            It may also take other arguments *args.
+        *args :
+            Additional non-stochastic arguments for func,
+            The function is computed as f(dstn, *args).
+
+        Returns
+        -------
+        f_dstn : DiscreteDistribution or DiscreteDistributionXRA
+            The distribution of func(dstn).
+        """
+        # we need to add one more dimension,
+        # the nature dimension, to any inputs that are n-dim arrays.
+        # This allows numpy to easily broadcast the function's output.
+        args = [
+            arg[..., np.newaxis] if isinstance(arg, np.ndarray) else arg for arg in args
+        ]
+        f_query = func(self.X, *args)
+
+        if xarray:
+            f_dstn = DiscreteDistributionXRA(
+                list(self.pmf), f_query, seed=self.seed, **kwargs
+            )
+        else:
+            f_dstn = DiscreteDistribution(list(self.pmf), f_query, seed=self.seed)
+
+        return f_dstn
+
+
+class DiscreteDistributionXRA(DiscreteDistribution):
+    """
+    A representation of a discrete probability distribution
+    stored in an underlying `xarray.DataArray` array.
+
+    Parameters
+    ----------
+    pmf : np.array
+        An array of floats representing a probability mass function.
+    data : np.array
+        Discrete point values for each probability mass.
+        For multivariate distributions, the last dimension of X must index
+        "nature" or the random realization. For instance, if X.shape == (2,6,4),
+        the random variable has 4 possible realizations and each of them has shape (2,6).
+    seed : int
+        Seed for random number generator.
+    coords : dict
+        Coordinate values/names for each dimension of the underlying array.
+    dims : tuple or list
+        Dimension names for each dimension of the underlying array.
```

Comment:
I'm sorry to be getting to a thorough review of this work so late.

What I am noticing here is that the arguments to this DiscreteDistributionXRA class are exposing the configuration of the underlying DataArray data structure: `dims`, `coords`. But the documentation is not providing guidance as to the semantics of these arguments with respect to the class's purpose as _a representation of a probability distribution._

Since you seem to be using these values mainly in the `labels` version of the `expected_value` operation, it leaves me with a lot of questions about the design here. I wonder if there's a clearer way to do what you're trying to do, or a way to describe this capability that doesn't rely so much on a deep understanding of the DataArray format.

It looks like the usage pattern you are intending here is for a multivariate distribution over A and B, you will have 'dims = ['rv', 'x']' where 'x' refers to... the probability mass? And then `coords= {'rv' : ['A','B']}`. The DataArray then stores a values indexed by ((a, b), x) which represent the probability of (A = a, B = b)... but then what is the 'x' dimension?

And if this is the intended usage pattern, would you ever want `dims` to be anything _other_ than `['rv', 'x']`? Would that be a reasonable default?

But I'm actually now quite confused about the data model. Why wouldn't you just have the `dims = ['A', 'B']`? That way each index of (a,b) would map to a probability value. Am I missing something?